client/task_runner.go

```diff
@@ -356,25 +423,287 @@ func (r *TaskRunner) validateTask() error {
+			case structs.VaultChangeModeSignal:
+				s, err := signals.Parse(r.task.Vault.ChangeSignal)
+				if err != nil {
+					e := fmt.Errorf("failed to parse signal: %v", err)
+					r.logger.Printf("[ERR] client: %v", err)
+					r.Kill("vault", e.Error())
+					return
+				}
+
+				if err := r.Signal("vault", "new Vault token acquired", s); err != nil {
+					r.logger.Printf("[ERR] client: failed to send signal to task %v for alloc %q: %v", r.task.Name, r.alloc.ID, err)
+					r.Kill("vault", fmt.Sprintf("failed to send signal to task: %v", err))
+					return
+				}
+			case structs.VaultChangeModeRestart:
+				r.Restart("vault", "new Vault token acquired")
+			case structs.VaultChangeModeNoop:
+				fallthrough
+			default:
+				r.logger.Printf("[ERR] client: Invalid Vault change mode: %q", r.task.Vault.ChangeMode)
+			}
+
+			// We have handled it
+			updatedToken = false
+
+			// Call the handler
+			r.updatedTokenHandler()
+		}
+
+		// Start watching for renewal errors
+		select {
+		case err := <-renewCh:
+			// Clear the token
+			token = ""
+			r.logger.Printf("[ERR] client: failed to renew Vault token for task %v on alloc %q: %v", r.task.Name, r.alloc.ID, err)
+
+			// Check if we have to do anything
+			if r.task.Vault.ChangeMode != structs.VaultChangeModeNoop {
+				updatedToken = true
+			}
+		case <-r.waitCh:
+			return
+		}
+	}
+}
+
+// deriveVaultToken derives the Vault token using exponential backoffs. It
+// returns the Vault token and whether the token is valid. If it is not valid we
+// are shutting down
+func (r *TaskRunner) deriveVaultToken() (string, bool) {
+	attempts := 0
+	for {
+		tokens, err := r.vaultClient.DeriveToken(r.alloc, []string{r.task.Name})
+		if err != nil {
+			backoff := (1 << (2 * uint64(attempts))) * vaultBackoffBaseline
+			if backoff > vaultBackoffLimit {
+				backoff = vaultBackoffLimit
+			}
+			r.logger.Printf("[ERR] client: failed to derive Vault token for task %v on alloc %q: %v; retrying in %v", r.task.Name, r.alloc.ID, err, backoff)
+
+			attempts++
+
+			// Wait till retrying
+			select {
+			case <-r.waitCh:
+				return "", false
+			case <-time.After(backoff):
+			}
+		} else {
+			return tokens[r.task.Name], true
+		}
+	}
+}
+
+// writeToken writes the given token to disk
+func (r *TaskRunner) writeToken(token string) error {
+	// Write the token to disk
+	secretDir, err := r.ctx.AllocDir.GetSecretDir(r.task.Name)
 	if err != nil {
-		err := fmt.Errorf("failed to build task's template manager: %v", err)
-		r.setState(structs.TaskStateDead, structs.NewTaskEvent(structs.TaskSetupFailure).SetSetupError(err))
-		r.logger.Printf("[ERR] client: alloc %q, task %q %v", r.alloc.ID, r.task.Name, err)
+		return fmt.Errorf("failed to determine task %s secret dir in alloc %q: %v", r.task.Name, r.alloc.ID, err)
+	}
+
+	// Write the token to the file system
+	tokenPath := filepath.Join(secretDir, vaultTokenFile)
+	if err := ioutil.WriteFile(tokenPath, []byte(token), 0777); err != nil {
```

Comment:
Are these permission safe because the parent dir isn't world readable (and thus the secret is only readable to root or within the chroot/container)?

[nit] `0666` might be more appropriate.